proposals/xxxx-experimental-dxil.md

```diff
@@ -0,0 +1,195 @@
+
+### Top 1 bit as "is experimental" flag
+
+The top bit of all opcodes is a flag stating if the opcode is experimental.
+
+No structural or shape changes to the DXIL occur, simply the fact that the opcode
+has the high bit set informs that it is experimental. This makes it very easy
+for the compiler and drivers to detect experimental opcodes. When an opcode is
+transistioned to stable the opcode needs to be assigned a stable number.
+This splits the 4 billion opcode space into two 2 billion partions. One for
+stable one for experimental.
+
+This is the simpliest proposal with the least invasive set of changes.
+
+Pros:
+ * Very simple
+ * Quick to implement
+ * Could be implemented "by hand" today by hard coding opcodes
+Cons:
+ * Not a solution for extensions
+ * transistion from experimental to stable isn't just unsetting the bit
+   * other stable ops may have already taken that number
+   * complicates the experimental->stable mapping
+
+### Top 8 bits as "opcode partition" value
+This is pretty much identical to the 1 bit flag proposal except there are 256
+partitions with 16 million opcodes each. The key difference is that it unlocks
+extension potential as extension developers such as IHVs could reserve a
+partition for their own use without collision with other opcodes.
+
+| Partition | Use |
+|-----------|-----|
+| 0 | stable |
+| 1 | experimental |
+| 2 | extension foo |
+| .. | extension .. |
+| 255 | extension 255 |
+
+
+Pros:
+ * Fairly simple
+ * Quick to implement
+ * Enables basic opcodes extension system
+Cons:
+ * transistion from experimental to stable isn't just clearing the partition
+   * other stable ops may have already taken that number
+   * complicates the experimental->stable mapping
+
+### Top 16 bits as "opcode partition" value
+Identical concept as above but with 64k partitions, each with 64k opcodes.
+
+### Split the opcode in half
+Lower 16 bits are the core/stable opcodes, Upper 16 bits are the experimental opcodes.
+
+Gives 64k opcodes for stable then the upper 64k can either be chunked manually
+leaving all number available for opcodes or it can be partitioned as 256
+chunks of 256 opcodes with the partition encoded into the opcode itself
+
+Very similar concept as before but keeping track of opcodes is complicated.
+Also enables a weird situation where two opcodes "could" be encoded into a
+single value.
+
+### Introduce dx.opx.opcodeclass for experimental/extended ops
+Denotes the experimental status in the actual opcode. Potentially doubles the
+opcode space depending on implementation however it doesn't make the transistion
+to stable any easier and complicated the integration with the current intrinsics
+system.
+
+Pros:
+ * Enables fairly robust extension system
+ * Doesn't consume large portions of the current opcode space
+ * obvious from reading the DXIL that experimental/extension is being used
+Cons:
+ * transistion from experimental to stable isn't just dropping the `x`
```

Comment:
This one actually makes the transition significantly harder than the other options. For the other options, you could use a regex search/replace to update opcodes (with some risk of missing some cases). In theory, you could even add a symbolic opcode replacer to avoid having to ever update the tests.

But with this approach, the instruction stream is changed in a way that can't easily be updated with text search/replace, or maped through a symbolic opcode replacement utility.

Reply:
Closing as has been discussed, added note about increased difficulty